[SIMD] SIMD functions should support Linear Scan and Graph Colouring register allocators.

justinmichaud opened the pull request:
#### 244d75cd9da4e7da65f1612fac1fbf240fa72104
<pre>
[SIMD] SIMD functions should support Linear Scan and Graph Colouring register allocators.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246348">https://bugs.webkit.org/show_bug.cgi?id=246348</a>

Reviewed by Yusuke Suzuki.

Today, we disable the linear scan and graph coloring register allocators
when WASM SIMD is enabled. Let&apos;s fix that by making them conservatively
treat floats as 128 bits when SIMD is enabled.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorNarrow):
(JSC::MacroAssemblerARM64::vectorMulSat):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::GenerateAndAllocateRegisters::generate):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackByLinearScan.cpp:
(JSC::B3::Air::allocateRegistersAndStackByLinearScan):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
(JSC::B3::Air::allocateRegistersByGraphColoring):
* Source/JavaScriptCore/b3/air/AirAllocateStackByGraphColoring.cpp:
(JSC::B3::Air::allocateStackByGraphColoring):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
(JSC::B3::Air::TmpWidth::recompute):
* Source/JavaScriptCore/jit/RegisterSet.h:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::recomputeDependentOptions):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/257519@main">https://commits.webkit.org/257519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e28b866cc3dde7652318c9812fb80b25b85c2f9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99205 "check-webkit-style running") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8417 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32338 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108598 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168851 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8969 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91710 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106535 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104961 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33775 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89913 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2297 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85744 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2188 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28862 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5169 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/42667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88608 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3666 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19827 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -557,7 +557,7 @@ bool canUseWebAssemblyFastMemory();
     v(Bool, useWebAssemblyTypedFunctionReferences, false, Normal, "Allow function types from the wasm typed function references spec.") \
     v(Bool, useWebAssemblyGC, false, Normal, "Allow gc types from the wasm gc proposal.") \
-    v(Bool, useWebAssemblySIMD, false, Normal, "Allow the new simd instructions and types from the wasm simd spec.") \
+    v(Bool, useWebAssemblySIMD, true, Normal, "Allow the new simd instructions and types from the wasm simd spec.") \
```
This plus the release asserts above are just for testing

Constellation commented on Source/JavaScriptCore/b3/air/AirRegLiveness.cpp:
```diff
@@ -31,13 +31,19 @@
 #include "AirInstInlines.h"
 
+namespace {
+    static constexpr bool debug = false;
```
We use variable name `verbose` for this purpose. Let's make it consistent with the other code.

Constellation commented on Source/JavaScriptCore/b3/air/AirRegLiveness.cpp:
```diff
@@ -31,13 +31,19 @@
 #include "AirArgInlines.h"
 #include "AirInstInlines.h"
 
+namespace {
```
Let's not use anonymous namespace, it is problematic with Unified Builds.
We are using FileName+Internal namespace, like, `namespace AirRegLivenessInternal`, and using `AirRegLivenessInternal::verbose`.

Constellation commented on Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
```diff
@@ -1709,13 +1709,13 @@ VectorMin U:G:Ptr, U:F:128, U:F:128, D:F:128
     SIMDInfo, Tmp, Tmp, Tmp
 
-VectorNarrow U:G:Ptr, U:F:128, U:F:128, D:F:128
+VectorNarrow U:G:Ptr, U:F:128, U:F:128, UD:F:128
```
I think U is not correct. MacroAssembler's effect model is not what the actually lowered instructions are doing.
U means that the value in `UD:F:128` affect on the result. But if it is not, regardless of how it is used, we should not attach U.

Constellation replied:
For example, ReplaceLane's destination's value still affects on the updated value. So it should be UD. But in this case, let's see vectorPmax implementation, destination value does not affect on the result, and it is always overridden by max value. Thus, it should be `D`.

justinmichaud replied:
Discussed offline: we will replace these with scratch for now, but ultimately we want to be able to say "early def"

Constellation commented on Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
```diff
@@ -1847,7 +1850,7 @@ arm64: VectorAddPairwise U:G:8, U:F:128, U:F:128, D:F:128
     SIMDInfo, Tmp, Tmp, Tmp
 
-VectorMulSat U:F:128, U:F:128, D:F:128
+VectorMulSat U:F:128, U:F:128, UD:F:128
```
`U` is not correct. VectorMulSat is also rewritten.

Constellation commented on Source/JavaScriptCore/b3/air/AirLivenessAdapter.h:
```diff
@@ -37,47 +37,53 @@
 
+namespace AirLivenessAdapter {
+static constexpr bool debug = false;
+}
```
Let's use `AirLivenessAdapterInternal` namespace. And use `verbose`.

Constellation commented on Source/JavaScriptCore/b3/air/AirLivenessAdapter.h:
```diff
@@ -37,47 +37,53 @@
     void prepareToCompute()
     {
+        if (AirLivenessAdapter::debug)
+            dataLogLn("Prepare to compute tmp or stack slot liveness for code: ", code);
```
Use `dataLogLnIf(AirLivenessAdapterInternal::verbose, "...")`

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
```diff
@@ -1739,7 +1758,7 @@ class ColoringAllocator : public AllocatorType<IndexType, InterferenceSet, bank>
 
             if (tmpWidth->defWidth(inst.args[0].tmp()) > Width32
-                && tmpWidth->useWidth(inst.args[1].tmp()) > Width32)
+                || tmpWidth->useWidth(inst.args[1].tmp()) > Width32) // OOPS fixme
```
OOPS is not supported format in WebKit. Always use FIXME.

justinmichaud replied:
Yeah, I wanted to make sure this wouldn't get landed. Is it correct that the merge-queue won't allow landing OOPS?

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackByLinearScan.cpp:
```diff
@@ -687,7 +689,7 @@ class LinearScan {
 void allocateRegistersAndStackByLinearScan(Code& code)
 {
-    RELEASE_ASSERT(!Options::useWebAssemblySIMD());
+    RELEASE_ASSERT(!Options::useWebAssemblySIMD()); // TEMP
```
Use FIXME.

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -488,6 +488,7 @@ void GenerateAndAllocateRegisters::prepareForGeneration()
 
 void GenerateAndAllocateRegisters::generate(CCallHelpers& jit)
 {
+    RELEASE_ASSERT(!Options::useWebAssemblySIMD()); // FIXME: temporary
```
Add a bit more description. What is temporary?

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -5293,12 +5299,13 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
-        vectorXor({ SIMDLane::v128, SIMDSignMode::None }, dest, dest, dest);
+        vectorMoveZero(dest);
         m_assembler.sqrdmlahv(dest, a, b, SIMDLane::i16x8);
     }
```
This change is no longer necessary.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -4979,6 +4979,11 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
+    void vectorMoveZero(FPRegisterID dest)
+    {
+        vectorXor({ SIMDLane::v128, SIMDSignMode::None }, dest, dest, dest);
+    }
```
Probably we should name it `moveZeroToVector`, aligned to `moveZeroToDouble`, `moveVector`, `moveDouble`.

Constellation reviewed (approved):
r=me

Commit:
[SIMD] SIMD functions should support Linear Scan and Graph Colouring register allocators.
https://bugs.webkit.org/show_bug.cgi?id=246348

Reviewed by Yusuke Suzuki.

Today, we disable the linear scan and graph coloring register allocators
when WASM SIMD is enabled. Let's fix that by making them conservatively
treat floats as 128 bits when SIMD is enabled.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorNarrow):
(JSC::MacroAssemblerARM64::vectorMulSat):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::GenerateAndAllocateRegisters::generate):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackByLinearScan.cpp:
(JSC::B3::Air::allocateRegistersAndStackByLinearScan):
* Source/JavaScriptCore/b3/air/AirAllocateRegistersByGraphColoring.cpp:
(JSC::B3::Air::allocateRegistersByGraphColoring):
* Source/JavaScriptCore/b3/air/AirAllocateStackByGraphColoring.cpp:
(JSC::B3::Air::allocateStackByGraphColoring):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
(JSC::B3::Air::TmpWidth::recompute):
* Source/JavaScriptCore/jit/RegisterSet.h:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::recomputeDependentOptions):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/257519@main
